cmd/juju/application/bundle/bundle.go

```diff
@@ -350,59 +345,74 @@ func verifyBundle(data *charm.BundleData, bundleDir string, verifyConstraints fu
 		for _, err := range verr.Errors {
 			errs = append(errs, err.Error())
 		}
+	}
+
+	if len(errs) > 0 {
 		return errors.New("the provided bundle has the following errors:\n" + strings.Join(errs, "\n"))
 	}
-	return errors.Trace(verifyError)
+	return nil
 }
 
-func verifyMixedSeriesBasesMatch(data *charm.BundleData) error {
-	if data == nil {
-		return nil
+type bundleSeriesData struct {
+	Series string `yaml:"series,omitempty"`
+
+	Applications map[string]struct {
+		Series string `yaml:"series"`
+	} `yaml:"applications"`
+
+	Machines map[string]struct {
+		Series string `yaml:"series"`
+	} `yaml:"machines"`
+}
+
+func verifyNoSeries(dslist []charm.BundleDataSource) []string {
+	errs := []string{}
+
+	// The first dslist is always the bundle itself
+	bundleErrs := verifyBundleNoSeries(dslist[0].RawBundle())
+	if len(bundleErrs) != 0 {
+		errs = append(errs, fmt.Sprintf("base bundle contains invalid key series:\n- %v", strings.Join(bundleErrs, "\n- ")))
 	}
-	if data.Series != "" && data.DefaultBase != "" {
-		b, err := corebase.ParseBaseFromString(data.DefaultBase)
-		if err != nil {
-			return errors.Trace(err)
-		}
-		s, err := corebase.GetSeriesFromBase(b)
-		if err != nil {
-			return errors.Trace(err)
-		}
-		if s != data.Series {
-			return errors.NewNotValid(nil, fmt.Sprintf("bundle series %q and base %q must match if both supplied", data.Series, data.DefaultBase))
+
+	// The rest are overlays
+	for i, ds := range dslist[1:] {
+		overlayErrs := verifyBundleNoSeries(ds.RawBundle())
+		if len(overlayErrs) != 0 {
+			errs = append(errs, fmt.Sprintf("overlay index %d contains invalid key series:\n- %v", i, strings.Join(overlayErrs, "\n- ")))
 		}
 	}
+	return errs
+}
 
-	for name, m := range data.Machines {
-		if m != nil && m.Series != "" && m.Base != "" {
-			b, err := corebase.ParseBaseFromString(m.Base)
-			if err != nil {
-				return errors.Trace(err)
-			}
-			s, err := corebase.GetSeriesFromBase(b)
-			if err != nil {
-				return errors.Trace(err)
-			}
-			if s != m.Series {
-				return errors.NewNotValid(nil, fmt.Sprintf("machine %q series %q and base %q must match if both supplied", name, m.Series, m.Base))
-			}
+func verifyBundleNoSeries(bundle string) []string {
+	dec := yaml.NewDecoder(strings.NewReader(bundle))
+
+	var errs []string
+	for docIdx := 0; ; docIdx++ {
+		var data *bundleSeriesData
+
+		err := dec.Decode(&data)
+		if err == io.EOF {
+			break
+		} else if err != nil {
+			// The bundle should already have been parsed if we're
+			// calling this check, so we should not see this
+			return []string{fmt.Sprintf("programming error: %w", err)}
```

Comment:
Remove `programming error`. I'd just return the error, things can be refactored and seeing `programming error` is not ideal.

Reply:
Done